test/Microsoft.AspNetCore.SignalR.Common.Tests/RedisCustomHubLifetimeManagerTests.cs

```diff
@@ -0,0 +1,484 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Internal;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR.Redis;
+using Microsoft.AspNetCore.SignalR.Tests;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Moq;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+using Xunit;
+
+namespace Microsoft.AspNetCore.SignalR.Common.Tests
+{
+    public class RedisCustomHubLifetimeManagerTests : HubLifeTimeManagerTestsBase<MyHub>
+    {
+        public override HubLifetimeManager<MyHub> GetNewHubLifetimeManager()
+        {
+            var server = new TestRedisServer();
+
+            return CreateLifetimeManager(server);
+        }
+
+        private RedisHubLifetimeManager<MyHub> CreateLifetimeManager(TestRedisServer server, MessagePackHubProtocolOptions messagePackOptions = null, JsonHubProtocolOptions jsonOptions = null)
+        {
+            var options = new RedisOptions() { ConnectionFactory = async (t) => await Task.FromResult(new TestConnectionMultiplexer(server)) };
+            messagePackOptions = messagePackOptions ?? new MessagePackHubProtocolOptions();
+            jsonOptions = jsonOptions ?? new JsonHubProtocolOptions();
+            return new RedisHubLifetimeManager<MyHub>(
+                NullLogger<RedisHubLifetimeManager<MyHub>>.Instance,
+                Options.Create(options),
+                new DefaultHubProtocolResolver(new IHubProtocol[]
+                {
+                    new JsonHubProtocol(Options.Create(jsonOptions)),
+                    new MessagePackHubProtocol(Options.Create(messagePackOptions)),
+                }, NullLogger<DefaultHubProtocolResolver>.Instance));
+        }
+
+        private async Task AssertMessageAsync(TestClient client)
+        {
+            var message = Assert.IsType<InvocationMessage>(await client.ReadAsync().OrTimeout());
+            Assert.Equal("Hello", message.Target);
+            Assert.Single(message.Arguments);
+            Assert.Equal("World", (string)message.Arguments[0]);
+        }
+        [Fact]
+        public async Task InvokeAllAsyncWithMultipleServersWritesToAllConnectionsOutput()
```

Comment:
What's redis specific about all of these tests?

Reply:
So there was actually a gap in the testing across the HubLifetimeManagers. Redis had a lot more tests than Default but they weren't all specific to the Redis scenario. Part of the clean up that I'm going to do.

Reply:
This shouldn't be redis specific, it should be scaleout specific. I think we should add another class that inherits from the generic lifetime manager test suite and provides scaleout specific tests.